### ISSUE TITLE
refresh delay for text input fields

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#1494]
With respect to [this comment](https://github.com/Jermolene/TiddlyWiki5/commit/2305385c08152c050946e4502735045a8e30aac1#commitcomment-9681386)...

For any widget implementing inputs of type **text** or **textarea**, implement a global delay only after which each input fires a refresh / some action, perhaps defaulting to somewhere between 200ms - 400ms, definitely non-zero.

This allows...
- typing in search boxes without TiddlyWiki being busy searching at each letter being typed
- typing in new markup without TiddlyWiki being busy rendering previews, even other page updates, e.g. lists
- etc...

Possibly not needed, perhaps, allow to override the global default with a **$delay** attribute for each widget.

Related discussion: https://groups.google.com/d/msg/tiddlywiki/hlIvXE6jRys/9ph05K1zAwAJ

**Comment by grayeul** [comment]
:+1:  I think that could be very helpful

**Comment by Jermolene** [comment]
There's already a somewhat similar mechanism that batches refresh cycles that only affect draft tiddlers. See the last bit of code here:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/render.js#L60

The key difference with this approach is that it still ensures that the tiddler store is up to date; it's just the widget tree that gets out of date, and that can always be refreshed. With the approach you suggest, we've got state in the widgets and some horrendous issues if the widget gets destroyed before the timeout elapses.

One brutal alternative proposal would be a namespace within which tiddlers have the behaviour that they only cause a refresh when their content goes over 'n' characters:

```
$:/state/MinimumLengthRefreshTiddlers/1/<mytiddler state suffix>
$:/state/MinimumLengthRefreshTiddlers/2/<mytiddler state suffix>
$:/state/MinimumLengthRefreshTiddlers/8/<mytiddler state suffix>
```

But that's really just a thought experiment; it's far too intrusive a solution to this problem.

**Comment by tobibeer** [comment]
Whatever the solution, TiddlyWiki hastily rushing into queries is quite (often) impinging on usability.

Yes, if there would be the case that a text widget gets destroyed before the input is handled, there needs to be a way to force-commit updates, e.g. on CTRL+ENTER when in edit-mode.

**Comment by Jermolene** [comment]
> Yes, if there would be the case that a text widget gets destroyed before the input is handled, there needs to be a way to force-commit updates, e.g. on CTRL+ENTER when in edit-mode.

Do you mean that the user would be required to use CTRL+ENTER to indicate that the value in the text box should be processed? The trouble is that it may be cleared by a refresh before the user gets a chance to type it.

**Comment by tobibeer** [comment]
> Do you mean that the user would be required to use CTRL+ENTER to indicate that the value in the text box should be processed?

I mean that there are events that immediately trigger that function to actually commit the value, e.g. CTRL+ENTER. However, a simple "change" or "keyup" (or whichever keyobard event) is not one of them, but rather delayed via setTimeout and then only calling that function if the source input still exists / hasn't already been commited... so, some flag indicates that commit-status.

> The trouble is that it may be cleared by a refresh before the user gets a chance to type it.

If that is so, then a user (or his wiki) must have performed some action that warrants destruction (and recreation).

**Comment by Jermolene** [comment]
> I mean that there are events that immediately trigger that function to actually commit the value, e.g. CTRL+ENTER. However, a simple "change" or "keyup" (or whichever keyboard event) is not one of them, but rather delayed via setTimeout and then only working that function if the source input still exists / hasn't already been commited.

But what if the user wants to use a mouse click to commit the value? I think this approach is fraught with difficulties, and suggest looking at the implementation I pointed to above that deals with minimising refreshes for draft tiddlers.

> If that is so, then a user (or his wiki) must have performed some action that warrants destruction (and recreation).

Not necessarily. Perhaps the sync mechanism just received an updated tiddler value from the server.

**Comment by tobibeer** [comment]
> But what if the user wants to use a mouse click to commit the value?

Then the value is immediately commited, setTimeout is only relevant to not disturb editing.


Draft tiddlers are really but one variant in all this and not necessarily the most important.

>  I think this approach is fraught with difficulties.

Not exactly sure which you mean precisely that wouldn't exist just the same without such a delay.

> Perhaps the sync mechanism just received an updated tiddler value from the server.

Then that update mechanism must not force-commit the update while there is an open handle on the same data. I don't know how to do such conflict handling, but with or without timeout, it's same-same... the data a user wishes to edit are changed under his nose, perhaps without them noticing.

**Comment by Jermolene** [comment]
@tobibeer I'm trying to say that I don't think that the proposal you've made is viable. We can't keep transient state in the widget tree, as we've now discussed many times. I'm pointing out the implementation of minimising updates to draft tiddlers not because I think your proposal has anything to do with draft tiddlers, but because it demonstrates a viable approach for controlling the refresh cycle for you to incorporate into a better proposal.

**Comment by tobibeer** [comment]
Mhhh, how is https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/render.js#L60 ...

* different from what I am proposing?
    * it seems we don't need other "transient states" than that which you implemented as "deferredChanges"
* handling all the events fired when keys are pressed in draft mode?
    * not sure I understand exactly how `timerId` actually works / is accessible across all calls to `$tw.wiki.addEventListener("change", ` ...what's it's lifetime? it feels like some rather well but weirdly scoped variable, that is accessible to the wiki object, but not global

**Comment by pmario** [comment]
>Draft tiddlers are but one variant in all this, not necessarily the most important. Above all, there are search / filter queries which can be immensely frustrating when triggered with every keypress. I have a cooking app and a dictionary app on my iphone both of which are a big pain precisely because of that... and it's not just that minimum amount of characters, it's doing the damn query at every key and refreshing the result list.

As I pointed out on a different topic. We have a performance problem on mobile devices. IMO most of it started to get "nasty" was when we introduced the flexible layout of the page template. ... There are way too many list widgets involved in the whole layout. ... 

Implementing timeouts here and there, imo is just fighting the symptoms and not the problem. 

... but, I don't have a solution :/

**Comment by tobibeer** [comment]
> Implementing timeouts here and there, imo is just fighting the symptoms and not the problem.

I don't quite understand the reasoning / alternatives (yet). If an action is not to immediately fire at every event, things like mousemove, what other basic choice do I have but to defer it via timeout?

**Comment by Jermolene** [comment]
>  how is https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/startup/render.js#L60 ... different from what I am proposing?

The difference is that you were proposing that the edit widget defer writing to the tiddler store, while the existing implementation doesn't defer writing to the store but instead defers the refresh cycle when it can. They are radically different approaches.

> handling all the events fired when keys are pressed in draft mode?
> not sure I understand exactly how timerId actually works / is accessible across all calls to $tw.wiki.addEventListener("change", ...what's it's lifetime? it feels like some rather well but weirdly scoped variable, that is accessible to the wiki object, but not global

Well, it's not working at the level of keystrokes, it works by checking to see if a tiddler change event involves only draft tiddlers, and if so it defers the refresh cycle.

`timerID` is just a regular local variable captured in a closure.

**Comment by tobibeer** [comment]
> The difference is that you were proposing that the edit widget defer writing to the tiddler store, while the existing implementation doesn't defer writing to the store but instead defers the refresh cycle when it can. They are radically different approaches.

I think it's slowly dawning on me. It's actually only the refresh at which that (search) query against the store is performed, e.g. in filters.

> Well, it's not working at the level of keystrokes, it works by checking to see if a tiddler change event involves only draft tiddlers, and if so it defers the refresh cycle.

That's what I meant, I don't quite understand how exactly an update would be deferred, especially with respect to any previous change having fired before. So far, I found keystrokes in the editor directly impacting on what I see in the UI, e.g. lists. (How) would one change event firing know of another one that is currently active ...with respect to `timerID`?

**Comment by Jermolene** [comment]
> That's what I meant, I don't quite understand how exactly an update would be deferred, especially with respect to any previous change having fired before. If they still follow suit in the same intervall, only delayed there wouldn't be any gain. So far, I found keystrokes in the editor directly impacting on what I see in the UI, e.g. lists. (How) would one change event firing know of another one that is currently active ...with respect to timerID?

The best advice I can give is to look at the code! It's set up to defer refreshes by 400ms, which you may find barely perceptible.

**Comment by tobibeer** [comment]
> The best advice I can give is to look at the code! It's set up to defer refreshes by 400ms, which you may find barely perceptible.

Ok, so I take it that **timerID** variable must be alive during successive calls so as to not do refreshes while there is one in the loop.

**Comment by Jermolene** [comment]
The timerID variable is captured in a closure so it is available to the inner functions. There's a good overview of closures here: http://stackoverflow.com/a/111200/3111487

**Comment by pmario** [comment]
@tobibeer ... may be OP is obsolete since we have the 3 char minimum length in for the text input fields. .. This can also be used for other stuff. ... so may be close this issue?

**Comment by tobibeer** [comment]
if i understood @Jermolene  right, this type of delaying the refresh is already implemented.
